File: packages/fintech/src/itch.js
// itch.js — NASDAQ TotalView-ITCH 5.0 flyweight parser + feed engine.
//
// Zero-allocation contract (docs/adapters/MANAGED-SEAMS-V1.md §2):
//   - One ItchView binds to the feed buffer once; `bind(offset)` re-points it.
//     No slices, no strings, no objects, no closures per message.
//   - All multi-byte fields are BIG-endian (network order) per the public
//     ITCH 5.0 spec. u48 timestamps reconstruct as hi*2**32+lo (exact: < 2^53).
//   - Order refs are exposed as lo/hi u32 PAIRS — never materialized as
//     doubles (refs may exceed 2^53; golden fixtures deliberately include
//     hi = 0x00C0FFEE refs).
//   - Unknown message types are SKIPPED by the u16 length prefix (counted).
//     Corrupt framing (length beyond buffer) fails closed with E_TRUNC.
//
// The view's semantic getters are TYPE-AWARE: message type is known at bind
// time, so `shares`/`price`/`match` resolve to the correct wire offsets for
// that message (e.g. replace shares at +27, trade match at +36). The book
// only ever sees semantic names.

export const E_TRUNC = 1;

export const T_SYSTEM = 0x53;      // 'S'
export const T_ADD = 0x41;         // 'A'
export const T_ADD_MPID = 0x46;    // 'F'
export const T_EXEC = 0x45;        // 'E'
export const T_EXEC_PRICE = 0x43;  // 'C'
export const T_CANCEL = 0x58;      // 'X'
export const T_DELETE = 0x44;      // 'D'
export const T_REPLACE = 0x55;     // 'U'
export const T_TRADE = 0x50;       // 'P'

export class ItchView {
  constructor(buffer) {
    this.dv = buffer instanceof DataView ? buffer : new DataView(buffer);
    this.off = 0;
    this.type = 0;
  }
  bind(off) {
    this.off = off;
    this.type = this.dv.getUint8(off);
    return this;
  }
  get locate() { return this.dv.getUint16(this.off + 1, false); }
  get tracking() { return this.dv.getUint16(this.off + 3, false); }
  get ts() {
    return this.dv.getUint16(this.off + 5, false) * 4294967296 +
      this.dv.getUint32(this.off + 7, false);
  }
  // u64 order ref as lo/hi u32 pair (BE wire order: hi at +11, lo at +15)
  get refLo() { return this.dv.getUint32(this.off + 15, false); }
  get refHi() { return this.dv.getUint32(this.off + 11, false); }
  // second ref (replace new-order id: hi at +19, lo at +23)
  get ref2Lo() { return this.dv.getUint32(this.off + 23, false); }
  get ref2Hi() { return this.dv.getUint32(this.off + 19, false); }
  get side() { return this.dv.getUint8(this.off + 19); } // 0x42 buy / 0x53 sell
  get shares() {
    const o = this.off;
    const t = this.type;
    // 'U' at +27; executed/cancelled shares ('E','C','X') at +19;
    // add/trade shares ('A','F','P') at +20
    if (t === T_REPLACE) return this.dv.getUint32(o + 27, false);
    if (t === T_EXEC || t === T_EXEC_PRICE || t === T_CANCEL)
      return this.dv.getUint32(o + 19, false);
    return this.dv.getUint32(o + 20, false);
  }
  get price() {
    const o = this.off;
    // 'A'/'F'/'P'/'C' at +32; 'U' at +31
    return this.type === T_REPLACE
      ? this.dv.getUint32(o + 31, false)
      : this.dv.getUint32(o + 32, false);
  }
  get match() {
    const o = this.off;
    if (this.type === T_TRADE) {
      return this.dv.getUint32(o + 36, false) * 4294967296 +
        this.dv.getUint32(o + 40, false);
    }
    return this.dv.getUint32(o + 23, false) * 4294967296 +
      this.dv.getUint32(o + 27, false);
  }
  eventCode() { return String.fromCharCode(this.dv.getUint8(this.off + 11)); }
  // Stock symbol as raw bytes into caller-owned scratch (no string alloc).
  stockInto(u8scratch) {
    const off = this.off + 24;
    for (let i = 0; i < 8; i++) u8scratch[i] = this.dv.getUint8(off + i);
    return u8scratch;
  }
}

// Walk framing and return the view-relative byte offset of each message
// START (payload start, after the 2-byte length prefix). Cold path.
export function frameOffsets(buffer, byteLength, maxCount = 1 << 20) {
  const dv = buffer instanceof DataView ? buffer : new DataView(buffer.buffer ?? buffer, buffer.byteOffset ?? 0, byteLength);
  const out = new Int32Array(maxCount);
  let off = 0;
  let n = 0;
  while (off + 2 <= byteLength && n < maxCount) {
    const len = dv.getUint16(off, false);
    out[n++] = off + 2;
    off += 2 + len;
  }
  return out.subarray(0, n);
}

// Feed engine: walks `[u16 BE len][payload]` framing, dispatches known
// messages into the book. One allocation total (the view), at construction.
export class ItchEngine {
  constructor(book, opts = {}) {
    this.book = book;
    this.view = new ItchView(new DataView(opts.buffer ?? new ArrayBuffer(64)));
    this.truncated = 0;
    this.bytesProcessed = 0;
    this.parseNs = 0; // accumulated parse+apply ns when a clock is supplied
    this._clock = opts.clock ?? null;
    this._resume = 0; // view-relative byte offset to resume from
  }

  // Returns E_TRUNC on corrupt framing, else 0. Counters live on the book.
  // `input` may be an ArrayBuffer, a DataView (windowed), or a typed-array
  // view (Buffer/subarray) — byte offsets are VIEW-RELATIVE. Processing
  // resumes from the previous end byte (engine keeps `_resume`), so a feed
  // can be pumped incrementally for checkpoint snapshots.
  process(input, endByte) {
    let dv;
    let len;
    if (input instanceof DataView) {
      dv = input;
      len = endByte ?? input.byteLength;
    } else if (input instanceof ArrayBuffer) {
      dv = new DataView(input);
      len = endByte ?? input.byteLength;
    } else {
      dv = new DataView(input.buffer, input.byteOffset, input.byteLength);
      len = endByte ?? input.byteLength;
    }
    const view = this.view;
    view.dv = dv;
    // Resume semantics apply ONLY to the same underlying buffer (incremental
    // checkpoint pumping over one stream). A new chunk buffer restarts at 0.
    if (this._lastBuffer !== dv.buffer || this._lastOffset !== dv.byteOffset) {
      this._resume = 0;
      this._lastBuffer = dv.buffer;
      this._lastOffset = dv.byteOffset;
    }
    const t0 = this._clock !== null ? this._clock() : 0;
    const book = this.book;
    let off = this._resume;
    let result = 0;
    while (off + 2 <= len) {
      const msgLen = dv.getUint16(off, false);
      if (off + 2 + msgLen > len) {
        this.truncated++;
        result = E_TRUNC;
        break;
      }
      view.bind(off + 2);
      book.applyView(view);
      this.bytesProcessed += 2 + msgLen;
      off += 2 + msgLen;
    }
    this._resume = off;
    if (this._clock !== null) this.parseNs += this._clock() - t0;
    return result;
  }
}

File: python/weft_fintech/src/weft_fintech/itch.py
# itch.py — NASDAQ TotalView-ITCH 5.0 flyweight parser (managed Python side).
#
# Mirrors packages/fintech/src/itch.js byte-for-byte:
#   - framing [u16 BE length][payload], unknown types skipped, E_TRUNC closed
#   - type-aware semantic fields (shares/price/match offsets per message type)
#   - u64 order refs handled as exact Python ints (never floats)
#   - u48 timestamps via hi * 2**32 + lo (exact)
#
# Zero-permanent-allocation contract: every per-message decode uses
# struct.Struct.unpack_from (a temporary tuple, freed by refcount before the
# next message) into a FLYWEIGHT — the view object is reused across the whole
# feed. tracemalloc proves EXACT 0 bytes of permanent growth over 1,000,000
# messages (tests/test_alloc.py):
#   - engine counters (bytes_processed / truncated / resume) live in ONE
#     preallocated array('q') — boxed int attributes would leave their FINAL
#     values as live blocks at measurement time;
#   - process() releases its decode scratch on return: the memoryview window
#     over the caller's buffer is re-pointed to the internal scratch created
#     in __init__, and the flyweight's lazily-cached u48/u64 fields are
#     reset to the -1 sentinel (small int, no heap block).
# Per-message fields are therefore only meaningful DURING the feed / inside
# apply callbacks — the same flyweight lifetime the TS engine documents.

import struct

from array import array

E_TRUNC = 1

T_SYSTEM = 0x53
T_ADD = 0x41
T_ADD_MPID = 0x46
T_EXEC = 0x45
T_EXEC_PRICE = 0x43
T_CANCEL = 0x58
T_DELETE = 0x44
T_REPLACE = 0x55
T_TRADE = 0x50

_U16 = struct.Struct('>H')
_U32 = struct.Struct('>I')


class ItchView:
    """Reusable flyweight over one ITCH message. bind(off) re-points it.

    Prefix fields (ts / ref / ref2) are lazily decoded once per bind and
    cached in __slots__ (sentinel -1: u48/u64 values are never negative),
    so hot messages only unpack the fields they actually use.
    """

    __slots__ = ('mv', 'off', 'type', '_ts', '_ref', '_ref2')

    def __init__(self, buffer):
        self.mv = memoryview(buffer)
        self.off = 0
        self.type = 0
        self._ts = -1
        self._ref = -1
        self._ref2 = -1

    def bind(self, off):
        self.off = off
        self.type = self.mv[off]
        self._ts = -1
        self._ref = -1
        self._ref2 = -1
        return self

    @property
    def locate(self):
        return _U16.unpack_from(self.mv, self.off + 1)[0]

    @property
    def tracking(self):
        return _U16.unpack_from(self.mv, self.off + 3)[0]

    @property
    def ts(self):
        v = self._ts
        if v < 0:
            hi = _U16.unpack_from(self.mv, self.off + 5)[0]
            lo = _U32.unpack_from(self.mv, self.off + 7)[0]
            v = hi * 4294967296 + lo
            self._ts = v
        return v

    # u64 order ref: hi at +11, lo at +15 (BE wire order)
    @property
    def ref(self):
        v = self._ref
        if v < 0:
            hi = _U32.unpack_from(self.mv, self.off + 11)[0]
            lo = _U32.unpack_from(self.mv, self.off + 15)[0]
            v = hi * 4294967296 + lo
            self._ref = v
        return v

    # second ref (replace new-order id): hi at +19, lo at +23
    @property
    def ref2(self):
        v = self._ref2
        if v < 0:
            hi = _U32.unpack_from(self.mv, self.off + 19)[0]
            lo = _U32.unpack_from(self.mv, self.off + 23)[0]
            v = hi * 4294967296 + lo
            self._ref2 = v
        return v

    @property
    def side(self):
        return self.mv[self.off + 19]  # 0x42 buy / 0x53 sell

    @property
    def shares(self):
        t = self.type
        off = self.off
        if t == T_REPLACE:                      # 'U' at +27
            return _U32.unpack_from(self.mv, off + 27)[0]
        if t == T_EXEC or t == T_EXEC_PRICE or t == T_CANCEL:  # 'E','C','X' at +19
            return _U32.unpack_from(self.mv, off + 19)[0]
        return _U32.unpack_from(self.mv, off + 20)[0]          # 'A','F','P'

    @property
    def price(self):
        t = self.type
        off = self.off
        if t == T_REPLACE:                      # 'U' at +31
            return _U32.unpack_from(self.mv, off + 31)[0]
        return _U32.unpack_from(self.mv, off + 32)[0]  # 'A','F','P','C'

    @property
    def match(self):
        t = self.type
        off = self.off
        if t == T_TRADE:                        # 'P' match at +36
            hi = _U32.unpack_from(self.mv, off + 36)[0]
            lo = _U32.unpack_from(self.mv, off + 40)[0]
        else:                                   # 'E','C' match at +23
            hi = _U32.unpack_from(self.mv, off + 23)[0]
            lo = _U32.unpack_from(self.mv, off + 27)[0]
        return hi * 4294967296 + lo

    def event_code(self):
        return chr(self.mv[self.off + 11])

    def stock_into(self, scratch):
        """Raw 8 stock bytes into a caller-owned writable (no string alloc)."""
        off = self.off + 24
        for i in range(8):
            scratch[i] = self.mv[off + i]
        return scratch


def frame_offsets(buffer, max_count=1 << 20):
    """View-relative payload-start offset of every framed message (cold)."""
    mv = memoryview(buffer)
    n = len(mv)
    out = []
    off = 0
    while off + 2 <= n and len(out) < max_count:
        (length,) = _U16.unpack_from(mv, off)
        out.append(off + 2)
        off += 2 + length
    return out


class ItchEngine:
    """Feed engine over a book. One view, reused. Resume-capable.

    Resume applies ONLY within one underlying buffer object (checkpoint
    pumping over one stream); a new chunk buffer restarts at 0 — same
    semantics as the TS engine.

    Zero-residue contract: counters are unboxed into _estats; decode scratch
    (buffer window + cached u48/u64 fields) is released when process()
    returns, so a traced run retains EXACTLY 0 bytes (Law 1 / Stage 3).
    """

    __slots__ = ('book', 'view', '_estats', '_last_buffer', '_scratch_mv')

    _E_TRUNCATED = 0
    _E_BYTES = 1
    _E_RESUME = 2

    def __init__(self, book):
        self.book = book
        scratch = bytearray(64)
        self.view = ItchView(scratch)
        self._scratch_mv = self.view.mv  # same object: restore target
        self._estats = array('q', bytes(8 * 3))
        self._last_buffer = None

    @property
    def truncated(self):
        return self._estats[ItchEngine._E_TRUNCATED]

    @property
    def bytes_processed(self):
        return self._estats[ItchEngine._E_BYTES]

    def process(self, buffer, end_byte=None):
        estats = self._estats
        if buffer is not self._last_buffer:
            estats[ItchEngine._E_RESUME] = 0
            self._last_buffer = buffer
        view = self.view
        mv = view.mv = memoryview(buffer)
        n = len(mv) if end_byte is None else end_byte
        view_bind = view.bind
        book_apply = self.book.apply_view
        off = estats[ItchEngine._E_RESUME]
        result = 0
        u16 = _U16.unpack_from
        trunc = ItchEngine._E_TRUNCATED
        bytes_e = ItchEngine._E_BYTES
        while off + 2 <= n:
            (msg_len,) = u16(mv, off)
            if off + 2 + msg_len > n:
                estats[trunc] += 1
                result = E_TRUNC
                break
            view_bind(off + 2)
            book_apply(view)
            estats[bytes_e] += 2 + msg_len
            off += 2 + msg_len
        estats[ItchEngine._E_RESUME] = off
        # release per-run decode scratch: next process() re-binds anyway
        view.mv = self._scratch_mv
        view.off = 0
        view.type = 0
        view._ts = -1
        view._ref = -1
        view._ref2 = -1
        return result
